### PR TITLE
Remote write sigv4

### LINKED ISSUE #3987: Sigv4 RemoteWrite Support
**What is missing?**

Prometheus 2.26.0 has added support for Sigv4 for the RemoteWrite configuration. (See [Changelog](https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md). This is not yet configurable through the related Prometheus-CDR.

**Why do we need it?**

Using Amazon Managed Service for Prometheus currently requires some sidecar container to sign every request during remote_write. This could be done through Prometheus directly now and would avoid additional configuration bloat.

**Environment**

* Prometheus Operator version:

    `539108b043e9ecc53c4e044083651e2ebfbd3492` (0.47.0 / 2021-04-13)

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by tolleiv**
This PR resolves #3987 and adds Sigv4 configuration support for the RemoteWriteSpec. Along with that it extends the related test with two different configuration scenarious.

**Release Note Template (will be copied)**

```release-note:feature 
Adds the ability to configure Sigv4 for RemoteWrite
```

### CODE REVIEW

File: pkg/prometheus/promcfg_test.go [hunk 1]
```diff
@@ -4151,6 +4153,13 @@ remote_write:
 			version: "v2.23.0",
```
Comment (paulfantom) [inline pkg/prometheus/promcfg_test.go:4153]: nit:
```suggestion
			version: "v2.26.0",
```

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -615,6 +617,22 @@ type QueueConfig struct {
...
+// Sigv4 optionally configures AWS's Signature Verification 4 signing process to
+// sign requests. Cannot be set at the same time as basic_auth or authorization.
+// +k8s:openapi-gen=true
+type Sigv4 struct {
+	// Region is the AWS region. If blank, the region from the default credentials chain used.
+	Region string `json:"region,omitempty"`
+	// AccessKey is the AWS API key. If blank, the environment variable `AWS_ACCESS_KEY_ID` is used
+	AccessKey string `json:"accessKey,omitempty"`
+	// SecretKey is the AWS API secret. If blank, the environment variable `AWS_SECRET_ACCESS_KEY` is used
+	SecretKey string `json:"secretKey,omitempty"`
```
Comment (s-urbaniak) [inline pkg/apis/monitoring/v1/types.go:629]: hm, both `AccessKey` and `SecretKey` should rather be of type `*v1.SecretKeySelector`.

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -1511,6 +1511,26 @@ func (cg *ConfigGenerator) generateRemoteWriteConfig(version semver.Version, p *
 			cfg = append(cfg, yaml.MapItem{Key: "proxy_url", Value: spec.ProxyURL})
 		}
 
+		if spec.Sigv4 != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1514]: we need to check the Prometheus version since it's only supported for versions >= v2.26.0.

**Comment by cabrinha** [comment]
Any updates on this PR?

**Comment by cabrinha** [comment]
@tolleiv any updates on this PR? Does it currently work?

**Comment by falmada** [comment]
Hey @tolleiv, anything we can do to help getting this closed? I'm trying to understand what is missing to get it added. 

Thanks!

**Comment by tolleiv** [comment]
Hi - sorry I got stuck with adding in the adjustments suggested by @s-urbaniak - using `*v1.SecretKeySelector` samewhat made sense but turned out very ugly. I didn't find time afterwards - sorry.

**Comment by jw-maynard** [comment]
@tolleiv are you giving up on this PR?

**Comment by mhausenblas** [comment]
@tolleiv is there anything I can do to unblock you?

**Comment by inge4pres** [comment]
Hey @tolleiv 😄 jumping on the bandwagon simply to add that if you don't have time to work on this PR during these days or ever firther, I am happy to carry forward your work, if you're willing to share access to your forked repo.

**Comment by tolleiv** [comment]
@inge4pres sent the invite as a collaborator - feel free to continue this

File: pkg/apis/monitoring/v1/types.go (hunk 2 above)
Comment (inge4pres) [reply pkg/apis/monitoring/v1/types.go:629]: Question on this: can we reuse the Prometheus type `SigV4Config` rather than rolling our own type?
https://github.com/prometheus/prometheus/blob/d5c3c567d3e2badbdcc844edeaeff4e3b36c71dc/config/config.go#L742

**Comment by inge4pres** [comment]
hey @s-urbaniak @simonpasquier @paulfantom I started looking into this: I am doing a rebase on recent `master` to include the sigV4 additions and I'd need a little guidance from you here.
Are the jsonnet files auto-generated: after running `make generate` I get a weird YAML output in bundle.yaml, with newlines that should not be there? 

```yaml
                    sigv4:
                      description: Sigv4 allows to configures AWS's Signature Verification
                        4
                      properties:
                        accessKey:
                          description: AccessKey is the AWS API key. If blank, the
                            environment variable `AWS_ACCESS_KEY_ID` is used
                          type: string
                        profile:
                          description: Profile is the named AWS profile used to authenticate.
                          type: string
                        region:
                          description: Region is the AWS region. If blank, the region
                            from the default credentials chain used.
                          type: string
                        roleArn:
                          description: RoleArn is the named AWS profile used to authenticate.
                          type: string
                        secretKey:
                          description: SecretKey is the AWS API secret. If blank,
                            the environment variable `AWS_SECRET_ACCESS_KEY` is used
```

Is there a rule that will split text after a certain number of characters?

**Comment by paulfantom** [comment]
> Are the jsonnet files auto-generated?

Yes. CRD files in jsonnet are generated from golang code by `scripts/generate-crds.go` script. You can regenerate them with `make generate-crds`.

> Is there a rule that will split text after a certain number of characters?

I believe this is embedded in controller-gen, we just follow what is set there. So yes, newlines in CRD YAML files are expected.

File: pkg/apis/monitoring/v1/types.go (hunk 2 above)
Comment (inge4pres) [reply pkg/apis/monitoring/v1/types.go:629]: Edited the comment above to have a better understanding of the initial request.

**Comment by inge4pres** [comment]
My only blocker is right now this https://github.com/prometheus-operator/prometheus-operator/pull/3994#discussion_r621970128

I looked deeper in the code base and found 2 possible ways of managing it:

* `store.Store` (user for AlertManager and other configurations' values from secrets)
* "custom loading" methods of `Operator` that will read from the list of secrets directly from the inClusterConfig client

Which one would you suggest to use for the AccessKey and SecretKey of SigV4 config?

**Comment by jw-maynard** [comment]
@inge4pres @paulfantom Any updates on this? Seems like we're waiting for feedback on how to store the secret information?

**Comment by inge4pres** [comment]
> @inge4pres @paulfantom Any updates on this? Seems like we're waiting for feedback on how to store the secret information?

Yeah, should I just propose a solution that I think it could fit and request re-review?

File: pkg/apis/monitoring/v1/types.go (hunk 2 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/types.go:629]: >  Does implies if they are stored in secrets, we should extract them from there.

yes you can have a look at GetSecretKey() for instance https://github.com/prometheus-operator/prometheus-operator/blob/63b3840645ce8ded070d1ceb7321ad19744a6644/pkg/assets/store.go#L263-L292

**Comment by simonpasquier** [comment]
@inge4pres I think that you need to extend `assets.Store` so it can deal with sigv4 credentials.

**Comment by inge4pres** [comment]
Hey there 😄 just pushed a revisioned SigV4 code after rebase on `master`.
There's one thing I was pondering while developing the code: it seems time for a bigger refactoring of `GenerateConfig` because now that `assets.Store` has been passed as a paramter, all the `additionalX` bytes slices can be inlined into it.
TBH I don't like a ot the final solution here, but I tried to stick with the rest of the codebase and, in the end,  found myself in a bit of a spaghetti situation.

The `assets.Store` is osed both from the caller of `GenerateConfig` and inside it, so there is an additional level of indirection that was not there before.

Note: I ran only `make test-unit` before submitting, I will setup all the env to run the full test suite locally if CI here does not pass.

**Comment by inge4pres** [comment]
Can someone please re-trigger CI actions? I amended a failed check on documentation 🙏🏼

File: pkg/prometheus/operator.go [hunk 4]
```diff
@@ -1589,7 +1589,7 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
 		additionalAlertRelabelConfigs,
 		additionalAlertManagerConfigs,
 		ruleConfigMapNames,
-	)
+		store)
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:1592]: to be consistent with basic auth, oauth and bearer tokens, we shouldn't have to pass the `store` variable to the config generator. Instead `assets.Store` should be extended to support sigv4 secrets and `GenerateConfig()` gets a new `store.Sigv4Assets` argument. Hope it makes sense :)

Comment (inge4pres) [reply pkg/prometheus/operator.go:1592]: That is kind of what I mentioned in https://github.com/prometheus-operator/prometheus-operator/pull/3994#issuecomment-872129578.

So in the intended solution, `store` will only be used to fetch a new `additionalSigV4Configs` as a `[]byte`? This would align the sigV4 config to others similar parameters passed?

Comment (simonpasquier) [reply pkg/prometheus/operator.go:1592]: Ah sorry i missed your initial comment! Currently the operator will collect the secret materials before synthesizing the configuration because it allows to skip service/pod monitors and probes that have invalid key selectors. Otherwise a "bad" resource could fail the reconciliation process.

So yes, the idea is that the `assets.Store` struct gets 2 new fields: 
* `SigV4AccessKeyAssets map[string]SigV4AccessKey` 
* `SigV4SecretKeyAssets map[string]SigV4SecretKey` 

https://github.com/prometheus-operator/prometheus-operator/blob/3f047bc9c74105ba774aff726798a66b81409f7f/pkg/assets/store.go#L40-L49

You need also 2 new methods (similar to AddBearerToken for instance):
* `AddSigV4AccessKey(ctx context.Context, ns string, sel v1.SecretKeySelector, key string) error` 
* `AddSigV4SecretKey(ctx context.Context, ns string, sel v1.SecretKeySelector, key string) error`

https://github.com/prometheus-operator/prometheus-operator/blob/3f047bc9c74105ba774aff726798a66b81409f7f/pkg/assets/store.go#L204-L218

And then call these functions whenever we do for `AddBearerToken` and others, for instance:

https://github.com/prometheus-operator/prometheus-operator/blob/3f047bc9c74105ba774aff726798a66b81409f7f/pkg/prometheus/operator.go#L1743-L1759

And the configuration generator function should be extended with 2 new arguments: `sigV4AccessKeyAssets map[string]store.SigV4AccessKey` and `sigV4SecretKeyAssets map[string]store.SigV4SecretKey`.

https://github.com/prometheus-operator/prometheus-operator/blob/3f047bc9c74105ba774aff726798a66b81409f7f/pkg/prometheus/promcfg.go#L159-L171

Does it make sense?

Side-note: #4180 refactors and simplifies the `GenerateConfig()` part in a nice way but it's not there yet so feel free to stick with what exists today.

**Comment by inge4pres** [comment]
hello 😄 
I addressed @simonpasquier [comment](https://github.com/prometheus-operator/prometheus-operator/pull/3994#discussion_r675660698) in latest push, didn't do _exactly_ what was suggested as I saw space for limiting the number of additional parameters to 1 (will simplify the refactoring in #4180 if this is merged before...), but the genral concepts have been applied.

I will try to setup e2e tests locally with `kind` so I don't have to wait for CI here to validate the changes, meanwhile could you please run it here to see if we are getting closer to merging this one?
Thanks 🙏🏼

**Comment by inge4pres** [comment]
I am rebasing this branch on top of most recent `master`, after #4180 is merged.
I see there was probably a missing communication, a misunderstanding which direction the codebase should take.

In this PR I proposed a similar change to [this one](https://github.com/prometheus-operator/prometheus-operator/pull/4180/files#r677327736), but I was asked to be consistent ([here](https://github.com/prometheus-operator/prometheus-operator/pull/3994#discussion_r674940715)) with the previous architecture.

@paulfantom @simonpasquier @s-urbaniak I understand this PR has been dragging for a while, and it is probably not a priority in the project's roadmap.
I will try again to make this change happen and I would like to ask for help in getting this through, in any form you might feel is the most appropriate.

**Comment by inge4pres** [comment]
> @inge4pres thank you for staying patient, the PR so far looks great  Let's fix two things:
> 
> 1. this needs a final rebase on master
> 2. You need to regenerate API docs as per the e2e failure:
> 
> ```
> --- a/Documentation/api.md
> +++ b/Documentation/api.md
> @@ -1052,11 +1052,14 @@ ServiceMonitorSpec contains specification parameters for a ServiceMonitor.
>  
>  Sigv4 optionally configures AWS's Signature Verification 4 signing process to sign requests. Cannot be set at the same time as basic_auth or authorization.
>  
> +
> +<em>appears in: [RemoteWriteSpec](#remotewritespec)</em>
> +
>  | Field | Description | Scheme | Required |
>  | ----- | ----------- | ------ | -------- |
>  | region | Region is the AWS region. If blank, the region from the default credentials chain used. | string | false |
> -| accessKey | AccessKey is the AWS API key. If blank, the default AWS SDK credentials are used. | *[v1.SecretKeySelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#secretkeyselector-v1-core) | false |
> -| secretKey | SecretKey is the AWS API secret. If blank, the default AWS SDK credentials are used. | *[v1.SecretKeySelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#secretkeyselector-v1-core) | false |
> +| accessKey | AccessKey is the AWS API key. If blank, the environment variable `AWS_ACCESS_KEY_ID` is used | *[v1.SecretKeySelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#secretkeyselector-v1-core) | false |
> +| secretKey | SecretKey is the AWS API secret. If blank, the environment variable `AWS_SECRET_ACCESS_KEY` is used | *[v1.SecretKeySelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#secretkeyselector-v1-core) | false |
>  | profile | Profile is the named AWS profile used to authenticate. | string | false |
>  | roleArn | RoleArn is the named AWS profile used to authenticate. | string | false |
> ```
> 
> Once done, I am fine with merging as all comments seem to be addressed

@s-urbaniak thanks for your help 🙏🏼 
I had spotted another E2E tests failure, due to timeout.
Unfortunately everyt single time  tried running `make test-e2e` on my local env I found that the wrong `kubectl` context was used for some tests and so they were timing out.

If you have a working local dev env, could you please run the E2E tests locally and validate they are ok?
I'll investigate a little further on the failure on my side.

**Comment by paulfantom** [comment]
@inge4pres Could you rebase and run `make Documentation/api.md`? Rebase might help with failing e2e tests and regenerating `api.md` will fix issue in `checks/Generate...` CI job.

**Comment by inge4pres** [comment]
> @inge4pres Could you rebase and run `make Documentation/api.md`? Rebase might help with failing e2e tests and regenerating `api.md` will fix issue in `checks/Generate...` CI job.

Hi,

sorry for the long delay in responding, I have been away from computers until today...
Thanks for the additional hint on the `make Documentation/api/md`, last time I tried simply editing manually that file, not knowing it was auto-generated.

I rebased, re-generated code and api.md running 
```
make Documentation/api.md generate
```

I have unit tests failing locally now, so no sense in running CI, I'll check what happened...

```
FAIL	github.com/prometheus-operator/prometheus-operator/pkg/prometheus [build failed]
```

**Comment by inge4pres** [comment]
I am running Go 1.17, is 1.16 required to run the tests locally?

File: go.mod [hunk 5]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/docker/distribution v2.7.1+incompatible
 	github.com/evanphx/json-patch/v5 v5.5.0
 	github.com/ghodss/yaml v1.0.0
+	github.com/go-kit/kit v0.10.0 // indirect
```
[the remark this answers is not in the document]
Comment (inge4pres) [reply go.mod:11]: Thanks! Just ran `make tidy`

File: pkg/prometheus/promcfg_test.go [hunk 6]
```diff
@@ -545,11 +541,11 @@ func TestProbeStaticTargetsConfigGenerationWithLabelEnforce(t *testing.T) {
 								"static":    "label",
 							},
 						},
-					},
-					MetricRelabelConfigs: []*monitoringv1.RelabelConfig{
-						{
-							Regex:  "noisy_labels.*",
-							Action: "labeldrop",
+						MetricRelabelConfigs: []*monitoringv1.RelabelConfig{
```
[the remark this answers is not in the document]
Comment (inge4pres) [reply pkg/prometheus/promcfg_test.go]: Indeed the build fail mentioned in https://github.com/prometheus-operator/prometheus-operator/pull/3994#issuecomment-910702941 is fixed now!

File: pkg/prometheus/promcfg_test.go [hunk 7]
```diff
@@ -4609,9 +4585,26 @@ remote_write:
 `,
 		},
 		{
-			version: "v2.23.0",
+			version: "v2.26.0",
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:4588]: could you create a new test case instead? this one is about testing that `MetadataConfig` is supported when version >= v2.23.0.

**Review by simonpasquier — COMMENTED** [review]
one nit otherwise lgtm!

File: pkg/prometheus/promcfg_test.go (hunk 7 above)
Comment (inge4pres) [reply pkg/prometheus/promcfg_test.go:4588]: If I read correctly, that test you are mentioning is below this one added after the rebase, at https://github.com/tolleiv/prometheus-operator/blob/remote_write_sigv4/pkg/prometheus/promcfg_test.go#L4659?

The diff doesn't help here 😄

Comment (inge4pres) [reply pkg/prometheus/promcfg_test.go]: Improved the diff in latest commit, I put the new sigv4 test at the end of `promcgf_test.go`.

**Comment by inge4pres** [comment]
I finally found the time to setup a test environment, and I possibly have found the problem spotted with timeouts in previous CI runs.

The Prometheus instance spawned by the CRD with the version in this branch, is continuously restarted by the operator.
I found these logs:

```
level=debug ts=2021-09-05T17:01:22.943011433Z caller=operator.go:662 component=alertmanageroperator msg="update handler" old=31348 cur=31349
level=debug ts=2021-09-05T17:01:22.9430524Z caller=operator.go:1130 component=prometheusoperator msg="update handler" old=31348 cur=31349
level=debug ts=2021-09-05T17:01:22.943160375Z caller=operator.go:562 component=alertmanageroperator msg="StatefulSet key did not match an Alertmanager key format" key=default/prometheus-prom-sigv4-kube-prometheus-prometheus
level=debug ts=2021-09-05T17:01:22.943186883Z caller=operator.go:517 component=thanosoperator msg="update handler" old=31348 cur=31349
level=debug ts=2021-09-05T17:01:22.943238126Z caller=operator.go:1140 component=prometheusoperator msg="StatefulSet updated"
level=debug ts=2021-09-05T17:01:22.961869817Z caller=operator.go:1130 component=prometheusoperator msg="update handler" old=31349 cur=31351
level=debug ts=2021-09-05T17:01:22.96191784Z caller=operator.go:1140 component=prometheusoperator msg="StatefulSet updated"
level=debug ts=2021-09-05T17:01:22.963131125Z caller=operator.go:662 component=alertmanageroperator msg="update handler" old=31349 cur=31351
level=debug ts=2021-09-05T17:01:22.963183559Z caller=operator.go:562 component=alertmanageroperator msg="StatefulSet key did not match an Alertmanager key format" key=default/prometheus-prom-sigv4-kube-prometheus-prometheus
level=debug ts=2021-09-05T17:01:22.963373564Z caller=operator.go:517 component=thanosoperator msg="update handler" old=31349 cur=31351
```

but I have no clue if they might be really the cause of the restart, they seem to be the only error though... any hints?

File: pkg/prometheus/promcfg_test.go [hunk 8]
```diff
@@ -4817,6 +4817,77 @@ remote_write:
   authorization:
     type: Bearer
     credentials: secret
+`,
+		},
+		{
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:4822]: Can you add a test case with an empty Sigv4 field (e.g. `Sigv4: {}`)?

**Review by simonpasquier — COMMENTED** [review]
I'm not sure about the restarts you're experiencing. I'd need to check locally with your PR.

File: pkg/prometheus/promcfg_test.go (hunk 8 above)
Comment (inge4pres) [reply pkg/prometheus/promcfg_test.go:4822]: Do you mean that no sigv4 configuration specified by the user will always result in an empty `sigv4: {}` in the Prometheus config file?
If yes, such a test would fail - should it pass?
Does Prometheus requires that section in the configuration file schema?
I see [here](https://github.com/prometheus/prometheus/blob/70f597b0334d48572c4b0c35940c975aef3031c7/docs/configuration/configuration.md#remote_write) that it is mutually exclusive with other authorizations methods.

Comment (simonpasquier) [reply pkg/prometheus/promcfg_test.go:4822]: The Prometheus docs say: `To use the default credentials from the AWS SDK, use sigv4: {}` and I suspect that most people would configure it like that. Hence it would be good to ensure that `spec: { remoteWrite: { sigv4: {}, ... } }` translates properly into the Prometheus config file.
If the sigv4 field is `nil` in the Prometheus resource then the the config file shouldn't contain the `sigv4` fiel.

>  I see here that it is mutually exclusive with other authorizations methods.

Correct. From what I can tell, we don't verify either that `Authorization` and `OAuth2` (for example) aren't specified at the same time but it would be good if you add a check for sigv4 at least.

Comment (inge4pres) [reply pkg/prometheus/promcfg_test.go:4822]: Rebased, added the sigv4 tests, and a small validation of the `RemoteWriteSpec` object(s) from the CRD.
I don't expect the final behavior to differ though? We did not change the way we check the configuration is in sync, and that seemed the problem causing the restart...

Comment (inge4pres) [reply pkg/prometheus/promcfg_test.go:4822]: Tested, again now: same problem as before - this seems to be the log from the offending piece of code that triggers the Pod restart
```
level=debug ts=2021-09-07T17:49:29.917839652Z caller=operator.go:1060 component=prometheusoperator msg="StatefulSet key did not match a Prometheus key format" key=default/alertmanager-prom-sigv4-kube-prometheus-alertmanager
```

**Comment by inge4pres** [comment]
> I'm not sure about the restarts you're experiencing. I'd need to check locally with your PR.

I am clueless as well 😄 
What I observe is the following: when the operator is started and the CRD does not include a `RemoteWrite.SigV`, the Prometheus instance is reconciled.
Then I add a secret holding the AWS access key and secret key to configure a SigV4.
As soon as the following SigV4 remote write section is added in the CRD, a restart loop of the Prometheus instance starts.

```
  remoteWrite:
  - name: amp
    url: https://aps-workspaces.eu-central-1.amazonaws.com/workspaces/ws-74b7801b-38f1-4737-9831-c4a56b48ff89/api/v1/remote_write
    remoteTimeout: "60"
    sigv4:
      region: eu-west-1
      accessKey:
        name: sigv4
        key: aws_access_key_id
      secretKey:
        name: sigv4
        key: aws_secret_access_key
```

I tried profiling the operator and I see that it spends 50% of its time between `pkg/prometheus.(*Operator).createOrUpdateRuleConfigMaps` (~30%) and `pkg/prometheus.(*Operator).createOrUpdateConfigurationSecret` (~20%), but I really can't figure out why 🤔

File: pkg/prometheus/operator.go [hunk 9]
```diff
@@ -2028,3 +2037,26 @@ func (c *Operator) listMatchingNamespaces(selector labels.Selector) ([]string, e
...
+		spec.Authorization,
+		spec.Sigv4}
+	var count int
+	for i, v := range mxFields {
+		if !reflect.ValueOf(v).IsNil() {
+			count++
+		}
+		if count > 1 {
+			return fmt.Errorf("invalid RemoteWriteSpec:\n%#v is not allowed together with\n%#v",
+				mxFields[i], mxFields[i-1])
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:2058]: this might not report the fields in conflict: if `BasicAuth` and `Sigv4` are both defined, it will log `Authorization` and `Sigv4`.

File: pkg/prometheus/promcfg_test.go (hunk 8 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg_test.go:4822]: I'll try to rebase and see if it helps

**Comment by simonpasquier** [comment]
in response to https://github.com/prometheus-operator/prometheus-operator/pull/3994#issuecomment-917442046, I've tried your example locally and I don't see that same issue (e.g. the reconciliation happens fine).
One note though: `remoteTimeout: "60"` should be `remoteTimeout: "60s"` otherwise Prometheus refuses to start.

As for why the CI times out for this pull request, I haven't figured out...

**Comment by simonpasquier** [comment]
@mhausenblas unfortunately not, I tried locally and everything works fine. I'm not even sure that the issue is with this change...

**Comment by mhausenblas** [comment]
Thanks a lot for the speedy reply @simonpasquier … what can we do to unblock @inge4pres?

**Comment by jeromeinsf** [comment]
I believe this stands for [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html)

**Comment by fpetkovski** [comment]
IRSA is natively implemented in the AWS SDK and the remote write functionality is part of Prometheus itself. The operator merely configures the Prometheus instance. Are you able to use IRSA with a Prometheus instance that's not managed by the operator?

Also for IRSA I believe you need to specify the IAM role that should be assumed.

**Comment by oussama-mechlaoui** [comment]
For more details about the context, I am using the [kube-prometheus](https://github.com/prometheus-operator/kube-prometheus) with jsonnet. Below the jsonnet configuration:

```
prometheus+: {
        spec+: {
          remoteWrite: [{
            url: 'https://aps-workspaces.eu-west-1.amazonaws.com/workspaces/****/api/v1/remote_write',
            sigv4: {
              region: 'eu-west-1',
            },
            queueConfig: {
              maxSamplesPerSend: 1000,
              maxShards: 200,
              capacity: 2500,
            },
          },
          ],
        },
      },
```

When generating the manifests, the prometheus CRD looks like:

```
  remoteWrite:
  - queueConfig:
      capacity: 2500
      maxSamplesPerSend: 1000
      maxShards: 200
    url: https://aps-workspaces.eu-west-1.amazonaws.com/workspaces/***/api/v1/remote_write
    sigv4:
      region: eu-west-1
```

However, when I check the prometheus server final config file below, I notice the sigv4 field is missing, hence, all ingest requests sent to Amazon Managed Prometheus were failing:

```
remote_write:
- url: https://aps-workspaces.eu-west-1.amazonaws.com/workspaces/****/api/v1/remote_write
  remote_timeout: 30s
  follow_redirects: true
  queue_config:
    capacity: 2500
    max_shards: 200
    min_shards: 1
    max_samples_per_send: 1000
    batch_send_deadline: 5s
    min_backoff: 30ms
    max_backoff: 100ms
  metadata_config:
    send: true
    send_interval: 1m
    max_samples_per_send: 500
```

PS: I have added accessKey and secretKey and they are reflected in the prometheus CRD, otherwise, the sigv4 field is always missing in the prometheus final config file.

**Comment by oussama-mechlaoui** [comment]
> IRSA is natively implemented in the AWS SDK and the remote write functionality is part of Prometheus itself. The operator merely configures the Prometheus instance. Are you able to use IRSA with a Prometheus instance that's not managed by the operator?
> 
> Also for IRSA I believe you need to specify the IAM role that should be assumed.

I have succeeded to configure sigv4 following steps in the [AWS Observability workshop](https://catalog.us-east-1.prod.workshops.aws/v2/workshops/31676d37-bbe9-4992-9cd1-ceae13c5116c/en-US/amp/ingest-metrics) in which prometheus server is not managed by prometheus-operator.

**Comment by fpetkovski** [comment]
Thanks @inge4pres and @oussama-mechlaoui for pointing those things out. This should be the test that @inge4pres is referring to: https://github.com/prometheus-operator/prometheus-operator/pull/3994/files#diff-40b00673aa9acb2e0dd5431cb36eedc1476124b086a3e070f5d6182709f8ea35R4953

@oussama-mechlaoui which prometheus version are you deploying? Can you please post the entire Prometheus resource spec?

**Comment by oussama-mechlaoui** [comment]
Prometheus version is 2.30.3. Below the generated prometheus CRD object:
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app.kubernetes.io/component: prometheus
    app.kubernetes.io/name: prometheus
    app.kubernetes.io/part-of: kube-prometheus
    app.kubernetes.io/version: 2.30.3
    prometheus: k8s
  name: k8s
  namespace: monitoring
spec:
  alerting:
    alertmanagers:
    - apiVersion: v2
      name: alertmanager-main
      namespace: monitoring
      port: web
  enableFeatures: []
  externalLabels: {}
  image: quay.io/prometheus/prometheus:v2.30.3
  nodeSelector:
    kubernetes.io/os: linux
  podMetadata:
    labels:
      app.kubernetes.io/component: prometheus
      app.kubernetes.io/name: prometheus
      app.kubernetes.io/part-of: kube-prometheus
      app.kubernetes.io/version: 2.30.3
  podMonitorNamespaceSelector: {}
  podMonitorSelector: {}
  probeNamespaceSelector: {}
  probeSelector: {}
  remoteWrite:
  - queueConfig:
      capacity: 2500
      maxSamplesPerSend: 1000
      maxShards: 200
    sigv4:
      region: eu-west-1
    url: https://aps-workspaces.eu-west-1.amazonaws.com/workspaces/****/api/v1/remote_write
  replicas: 2
  resources:
    requests:
      memory: 400Mi
  ruleNamespaceSelector: {}
  ruleSelector: {}
  securityContext:
    fsGroup: 2000
    runAsNonRoot: true
    runAsUser: 1000
  serviceAccountName: prometheus-k8s
  serviceMonitorNamespaceSelector: {}
  serviceMonitorSelector: {}
  version: 2.30.3
```

**Comment by pr0PM** [comment]
Using the newer version of AMP and updated charts for prometheus-operator via kubePrometheusStack and facing this error. Looks like some sigv4 issue, anyone faced this?

```
prometheus ts=2021-11-29T13:43 caller=dedupe.go:112 component=remote level=error remote_name=816243 url=https://workspaces.us-east-1.amazonaws.com/workspaces/ws-redactedapi/v1/remote_write msg="non-recoverable error" count=146 exemplarCount=0 err="server returned HTTP status 403 Forbidden: {\"message\":\"Missing Authentication Token\"}"
```
